File: clients/android/app/src/main/kotlin/com/lextures/android/features/livequiz/LiveGameSocket.kt
package com.lextures.android.features.livequiz

import com.lextures.android.core.config.AppConfiguration
import com.lextures.android.core.lms.LiveGameAnswerAck
import com.lextures.android.core.lms.LiveGameInboundMessage
import com.lextures.android.core.lms.LiveGameLogic
import com.lextures.android.core.lms.LiveGameStateFrame
import com.lextures.android.core.lms.LiveQuizLogic
import com.lextures.android.core.lms.LiveQuizObservability
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import org.json.JSONObject
import java.time.Instant

/**
 * Live-quiz WebSocket with role + playerToken handshake (IQ.3 / MOB.5).
 * Dedicated client — shared [com.lextures.android.core.realtime.WebSocketClient] only sends `{authToken}`.
 */
class LiveGameSocket(
    private val courseCode: String,
    private val gameId: String,
    private val role: LiveGameLogic.Role = LiveGameLogic.Role.Player,
    private val playerToken: String?,
    private val accessTokenProvider: () -> String?,
    private val onState: (LiveGameStateFrame) -> Unit,
    private val onAck: (LiveGameAnswerAck) -> Unit,
    private val onKicked: () -> Unit,
    private val onConn: (LiveGameLogic.ConnStatus) -> Unit,
    private val http: OkHttpClient = OkHttpClient(),
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
    private var socket: WebSocket? = null
    private var isStopped = true
    private var reconnectJob: Job? = null
    private var retry = 0
    private var seq = 0
    private var kicked = false

    fun connect() {
        if (kicked) return
        isStopped = false
        reconnectJob?.cancel()
        reconnectJob = null
        val auth = accessTokenProvider()?.takeIf { it.isNotBlank() }
        if (auth == null && playerToken.isNullOrBlank()) {
            scheduleReconnect()
            return
        }
        openConnection(auth)
    }

    fun disconnect() {
        isStopped = true
        reconnectJob?.cancel()
        reconnectJob = null
        socket?.close(1000, null)
        socket = null
    }

    fun send(payload: JSONObject) {
        socket?.send(payload.toString())
    }

    fun submitAnswer(
        questionIndex: Int,
        answer: LiveGameLogic.AnswerPayload,
        powerUp: String? = null,
    ) {
        send(
            LiveGameLogic.answerMessage(
                questionIndex = questionIndex,
                answer = answer,
                clientSentAt = Instant.now().toString(),
                powerUp = powerUp,
            ),
        )
    }

    private fun openConnection(authToken: String?) {
        socket?.close(1000, null)
        onConn(if (retry > 0) LiveGameLogic.ConnStatus.Reconnecting else LiveGameLogic.ConnStatus.Connecting)

        val path = LiveQuizLogic.webSocketPath(courseCode, gameId)
        val url = AppConfiguration.apiUrl(path).toString()
            .replaceFirst("https://", "wss://")
            .replaceFirst("http://", "ws://")
        val request = Request.Builder().url(url).build()

        socket = http.newWebSocket(
            request,
            object : WebSocketListener() {
                override fun onOpen(webSocket: WebSocket, response: Response) {
                    webSocket.send(
                        LiveGameLogic.authHandshake(authToken, role, playerToken).toString(),
                    )
                    if (seq > 0) {
                        webSocket.send(LiveGameLogic.catchupMessage(seq).toString())
                    } else if (role == LiveGameLogic.Role.Player) {
                        webSocket.send(LiveGameLogic.helloMessage(0).toString())
                    }
                }

                override fun onMessage(webSocket: WebSocket, text: String) {
                    when (val msg = LiveGameLogic.parseInbound(text)) {
                        LiveGameInboundMessage.Kicked -> {
                            kicked = true
                            isStopped = true
                            onKicked()
                            onConn(LiveGameLogic.ConnStatus.Kicked)
                            webSocket.close(1000, null)
                        }
                        is LiveGameInboundMessage.AnswerAck -> onAck(msg.ack)
                        is LiveGameInboundMessage.State -> {
                            seq = msg.frame.seq
                            onState(msg.frame)
                            onConn(
                                if (msg.frame.phase == LiveGameLogic.Phase.Ended.wire) {
                                    LiveGameLogic.ConnStatus.Ended
                                } else {
                                    LiveGameLogic.ConnStatus.Connected
                                },
                            )
                            retry = 0
                            LiveQuizObservability.record("live_quiz_reconnect", mapOf("ok" to "1"))
                        }
                        LiveGameInboundMessage.Unknown -> Unit
                    }
                }

                override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                    handleDisconnect(webSocket)
                }

                override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                    handleDisconnect(webSocket)
                }
            },
        )
    }

    private fun handleDisconnect(finished: WebSocket) {
        if (isStopped || kicked || socket !== finished) return
        socket = null
        retry += 1
        LiveQuizObservability.record("live_quiz_reconnect", mapOf("ok" to "0"))
        scheduleReconnect()
    }

    private fun scheduleReconnect() {
        onConn(LiveGameLogic.ConnStatus.Reconnecting)
        val delayMs = LiveGameLogic.reconnectDelayMs(retry).toLong()
        reconnectJob = scope.launch {
            delay(delayMs)
            connect()
        }
    }
}

File: clients/ios/Lextures/Features/LiveQuiz/LiveGameSocket.swift
import Foundation

/// Live-quiz WebSocket with role + playerToken handshake (IQ.3 / MOB.5).
/// Dedicated client — shared `WebSocketClient` only sends `{authToken}`.
@MainActor
final class LiveGameSocket {
    private let courseCode: String
    private let gameId: String
    private let role: LiveGameLogic.Role
    private let playerToken: String?
    private let accessTokenProvider: () -> String?
    private let onState: (LiveGameStateFrame) -> Void
    private let onAck: (LiveGameAnswerAck) -> Void
    private let onKicked: () -> Void
    private let onConn: (LiveGameLogic.ConnStatus) -> Void

    private var task: URLSessionWebSocketTask?
    private var isStopped = true
    private var reconnectTask: Task<Void, Never>?
    private var retry = 0
    private var seq = 0
    private var kicked = false

    init(
        courseCode: String,
        gameId: String,
        role: LiveGameLogic.Role = .player,
        playerToken: String?,
        accessTokenProvider: @escaping () -> String?,
        onState: @escaping (LiveGameStateFrame) -> Void,
        onAck: @escaping (LiveGameAnswerAck) -> Void,
        onKicked: @escaping () -> Void,
        onConn: @escaping (LiveGameLogic.ConnStatus) -> Void
    ) {
        self.courseCode = courseCode
        self.gameId = gameId
        self.role = role
        self.playerToken = playerToken
        self.accessTokenProvider = accessTokenProvider
        self.onState = onState
        self.onAck = onAck
        self.onKicked = onKicked
        self.onConn = onConn
    }

    func connect() {
        guard !kicked else { return }
        isStopped = false
        reconnectTask?.cancel()
        reconnectTask = nil
        let auth = accessTokenProvider()
        if (auth == nil || auth?.isEmpty == true) && (playerToken == nil || playerToken?.isEmpty == true) {
            scheduleReconnect()
            return
        }
        openConnection(authToken: auth)
    }

    func disconnect() {
        isStopped = true
        reconnectTask?.cancel()
        reconnectTask = nil
        task?.cancel(with: .normalClosure, reason: nil)
        task = nil
    }

    func send(_ object: [String: Any]) {
        guard let task, let data = try? JSONSerialization.data(withJSONObject: object),
              let text = String(data: data, encoding: .utf8) else { return }
        Task {
            try? await task.send(.string(text))
        }
    }

    func submitAnswer(questionIndex: Int, answer: LiveGameLogic.AnswerPayload, powerUp: String? = nil) {
        let iso = ISO8601DateFormatter().string(from: Date())
        send(LiveGameLogic.answerMessage(
            questionIndex: questionIndex,
            answer: answer,
            clientSentAt: iso,
            powerUp: powerUp
        ))
    }

    private func openConnection(authToken: String?) {
        task?.cancel(with: .normalClosure, reason: nil)
        onConn(retry > 0 ? .reconnecting : .connecting)

        let path = LiveQuizLogic.webSocketPath(courseCode: courseCode, gameId: gameId)
        let newTask = WebSocketSessionSupport.shared.session.webSocketTask(
            with: AppConfiguration.webSocketURL(path: path)
        )
        task = newTask
        newTask.resume()

        Task { [weak self] in
            guard let self else { return }
            let opened = await WebSocketSessionSupport.shared.waitForOpen(newTask)
            guard opened, !self.isStopped, self.task === newTask else {
                await self.handleDisconnect()
                return
            }
            let handshake = LiveGameLogic.authHandshake(
                authToken: authToken,
                role: self.role,
                playerToken: self.playerToken
            )
            if let data = try? JSONSerialization.data(withJSONObject: handshake),
               let text = String(data: data, encoding: .utf8) {
                try? await newTask.send(.string(text))
            }
            if self.seq > 0 {
                self.send(LiveGameLogic.catchupMessage(afterSeq: self.seq))
            } else if self.role == .player {
                self.send(LiveGameLogic.helloMessage(resumeSeq: 0))
            }
            do {
                try await self.receiveLoop(task: newTask)
            } catch {
                await self.handleDisconnect()
            }
        }
    }

    private func receiveLoop(task: URLSessionWebSocketTask) async throws {
        while true {
            let message = try await task.receive()
            let data: Data?
            switch message {
            case .data(let d): data = d
            case .string(let text): data = text.data(using: .utf8)
            @unknown default: data = nil
            }
            guard let data else { continue }
            switch LiveGameLogic.parseInbound(data) {
            case .kicked:
                kicked = true
                isStopped = true
                onKicked()
                onConn(.kicked)
                task.cancel(with: .normalClosure, reason: nil)
                return
            case .answerAck(let ack):
                onAck(ack)
            case .state(let frame):
                seq = frame.seq
                onState(frame)
                onConn(frame.phase == LiveGameLogic.Phase.ended.rawValue ? .ended : .connected)
                retry = 0
                LiveQuizObservability.record("live_quiz_reconnect", attributes: ["ok": "1"])
            case .unknown:
                break
            }
        }
    }

    private func handleDisconnect() async {
        guard !isStopped, !kicked else { return }
        task = nil
        retry += 1
        LiveQuizObservability.record("live_quiz_reconnect", attributes: ["ok": "0"])
        scheduleReconnect()
    }

    private func scheduleReconnect() {
        onConn(.reconnecting)
        let delayMs = LiveGameLogic.reconnectDelayMs(retry: retry)
        reconnectTask = Task { [weak self] in
            try? await Task.sleep(for: .milliseconds(delayMs))
            guard let self, !Task.isCancelled else { return }
            self.connect()
        }
    }
}
